Add email form at bug report page

bdmendes opened the pull request:
[Closes] #302 
User can now include their email in a bug report. A hint is provided that the info will be public on GitHub when the user touches the box.

![email-bug-report](https://user-images.githubusercontent.com/61701401/107945795-0c728800-6f88-11eb-8e00-72f70cdfde8a.jpg)


# Review checklist
- [x] Terms and conditions reflect the current change
- [x] Contains enough appropriate tests
- [ ] Increments version in `pubspec.yaml` (changes `1.0.0+1` to `1.0.0+2` for example)
- [x] If PR includes UI updates/additions, its description has screenshots
- [x] Behavior is as expected
- [x] Clean, well structured code

reeckset commented on app_feup/ios/Flutter/flutter_export_environment.sh:
```diff
@@ -1,10 +1,10 @@
+export "FLUTTER_FRAMEWORK_DIR=/opt/flutter/bin/cache/artifacts/engine/ios"
 export "FLUTTER_BUILD_NAME=1.0.0"
-export "FLUTTER_BUILD_NUMBER=1"
+export "FLUTTER_BUILD_NUMBER=3"
```
We should add this file to .gitignore and remove its changes from this PR

bdmendes replied:
It does seem to be in the gitignore of the ios folder, not sure how it got tracked

reeckset commented on app_feup/lib/view/Widgets/bug_report_form.dart:
```diff
@@ -91,16 +88,27 @@ class BugReportFormState extends State<BugReportForm> {
+      maxLines: 2,
+      description: 'Contacto (opcional)',
+      labelText: 'Email em que desejas ser contactado',
+      hintText: 'Informação pública na página do GitHub!',
```
Maybe we should make this more explicit to prevent any legal issues or misunderstandings. Can we add a dialog box upon submission of the bug (whether the user provided email address or not) saying something like "A informação submetida será disponibilizada publicamente na plataforma GitHub" with two buttons "Cancelar" and "Aceito"?

bdmendes replied:
Maybe it would be even simpler to have that text in a checkbox above the send button, and disallow sending if the checkbox isn't checked, with an error message?

reeckset replied:
That would work also yes. If it doesn't clutter the screen too much, it works for me!

reeckset reviewed (commented):
Nice! Very useful for the dev team indeed. Looks great, but pointed out two small things

bdmendes commented:
Added a consent box.
To make the email form actually optional, added a parameter in the FormTextField class to relax the validator if needed.

![email-bug-report](https://user-images.githubusercontent.com/61701401/107973531-38eccb00-6fad-11eb-9d7e-e5f63ebff7d3.jpg)

andrefmrocha commented on app_feup/lib/view/Widgets/bug_report_form.dart:
```diff
@@ -168,19 +178,47 @@ class BugReportFormState extends State<BugReportForm> {
         'Enviar',
         style: TextStyle(color: Colors.white, fontSize: 20.0),
       ),
+      disabledColor: Theme.of(context).disabledColor,
+      disabledElevation: 1,
```
Shouldn't the `isConsentGiven` be passed to the `enabled` field of the button? https://api.flutter.dev/flutter/material/MaterialButton/enabled.html

bdmendes replied:
I think that field is not accessible, from what I've seen here: https://api.flutter.dev/flutter/material/RaisedButton-class.html
The button is disabled when onPressed is null

andrefmrocha reviewed (changes requested):
Good work! Just two small things

bdmendes commented on app_feup/lib/view/Widgets/bug_report_form.dart:
```diff
@@ -168,19 +178,47 @@ class BugReportFormState extends State<BugReportForm> {
-          submitBugReport();
-        }
-      },
+      onPressed: !_isConsentGiven
+          ? null
+          : () {
+              if (_formKey.currentState.validate() && !_isButtonTapped) {
+                submitBugReport();
+              }
+            },
```
I did try returning null inside the function, but it seems that the actual function must be null for the button to be disabled

andrefmrocha commented:
Alright this sounds good to go, but can you please increment the application version the `pubspec.yaml`

andrefmrocha reviewed (dismissed):
Great work :rocket:

andrefmrocha commented:
Can you approve and merge @reeckset ?

bdmendes commented:
Rebased with upstream

reeckset commented on app_feup/lib/view/Widgets/bug_report_form.dart:
```diff
@@ -168,19 +178,47 @@ class BugReportFormState extends State<BugReportForm> {
+        child: CheckboxListTile(
+          activeColor: Theme.of(context).primaryColor,
+          title: Text(
+              '''Consinto que a informação seja publicada no GitHub, incluindo o meu contacto pessoal, se fornecido.''',
```
Can it be more explicit that it's public? Something like "Consinto que toda esta informação será disponibilizada publicamente na plataforma GitHub, incluindo o meu contacto pessoal, se fornecido". Most importantly, we have to follow the GDPR rules I mentioned above.

bdmendes replied:
Sure

reeckset commented on app_feup/lib/view/Widgets/bug_report_form.dart:
```diff
@@ -91,16 +89,29 @@ class BugReportFormState extends State<BugReportForm> {
     ));
 
+    formWidget.add(FormTextField(
+      emailController,
```
Should we add email validation here?

bdmendes replied:
I think a simple regex validator would do. Should I extend the FormFieldClass to override the validator or perhaps include as a parameter in the existing class?

reeckset replied:
I think there's already support for validators, but using a TextFormField instead of a FormTextField? I don't get it either ahah. Take a look at this: https://stackoverflow.com/a/63293006

bdmendes replied:
My question was what's the best design, pass validateEmail as a new parameter to FormTextField and check if it's set in the validator or extend FormTextField to something like EmailFormTextField to override the validator. Also, would it be worth to add a new package, giving the regex check is two lines (defining and checking for a match)?

bdmendes replied:
Regarding the last question, I took a look at the source code of `email_validator`and it is indeed small, I'm including it

reeckset replied:
Yes using that package should not be a problem, but my suggestion was regarding the design too. It seems that the **TextFormField** they use already has a "validator" parameter which makes it easier to maintain and I believe also gives [UI feedback to the user](https://api.flutter.dev/flutter/material/TextFormField-class.html). If you didn't use that package you could create your own validator separately and use it there, but it's already very small so why not xD

bdmendes replied:
In the entire bug report page we're using a custom `FormTextField` class which wraps `TextFormField` from the Material Design widgets. This is confusing naming schema to me, however I won't change it in the scope of this PR. This does imply however that the interface is a bit different and the validator behaviour depends on some parameters. To make the validator work for emails I added a new attribute. Please check it out

reeckset reviewed (changes requested):
The technical aspect is great. I like how the UX turned out. But...

A GDPR thought I am having is:
- We have to inform users that this data will be kept indefinitely
- We have to give them a way to erase this data (by contacting our email for example) - this might be hard because github doesn't allow deleting issues easily as far as I know.

So... Should move this system to another platform and make it private (or at least easy to delete)? @andrefmrocha

andrefmrocha commented:
> So... Should move this system to another platform and make it private (or at least easy to delete)? @andrefmrocha

I think a good idea for the time being would be to create a private repository for the issues of our users. Whenever something was deemed relevant, they would be published in our issues as an anonymous message. Wdyt?

andrefmrocha commented on app_feup/lib/view/Widgets/form_text_field.dart:
```diff
@@ -57,9 +64,14 @@ class FormTextField extends StatelessWidget {
-                  return emptyText;
+                  return isOptional ? null : emptyText;
+                }
+                switch (validatorType) {
+                  case FormTextFieldValidator.email:
+                    return EmailValidator.validate(value) ? null : emptyText;
+                  default:
+                    return null;
```
I think it would be best if the validator function itself was passed to the argument and this function would just become a check if the function is present or not. Wdyt?

bdmendes replied:
I'm thinking of a hybrid solution, maybe use a default validator that handles the optional situation, and make it call another function that handles the formatting stuff which was passed by argument (defaults to null). Wdyt?

andrefmrocha replied:
Sounds good!

bdmendes commented:
Can't we simply login with the bot account and edit the issue of post if a user requests that? Is there a time limit for that?

andrefmrocha commented:
> Can't we simply login with the bot account and edit the issue of post if a user requests that? Is there a time limit for that? 

Editing a message does not delete the previous, there is a edition history on every message

bdmendes commented:
See #342
